05_testing_and_ci/python_testing_demo.md

```diff
@@ -0,0 +1,58 @@
+# Notes for Demos of Python Testing Frameworks
+
+Example code is in [05_testing_and_ci/examples](https://github.com/Simulation-Software-Engineering/Lecture-Material/blob/main/05_testing_and_ci/examples)
+
+## Software code used
+
+- The file `operations.py` consists of two functions `find_max` and `find_average` which calculate the maximum and average of elements of a list. The `main()` routine in the file applies the functions to a list and prints the output.
+- `main()` function in `operations.py` has assertion statements to check if the correct data type is passed to specific functions.
+- Assertion statements are the most basic way of testing code and are also used in unit and integration testing as demonstrated here.
+- Tests are written in the file `test_operations.py`. The `test_*` prefix in the name is required so that pytest detects the file as a testing file. Suffix form `*_test.py` also works.
+- In all there are two unit tests, one integration test and one regression test.
+- The unit tests test the individual functions `find_max` and `find_average`.
+- The integration test triggers both the functions `find_max` and `find_average` and checks that the average is less than the maximum, something that should always be true for a set of numbers.
+- The regression test first reads an old data set and an average value from a CSV file. Then the function `find_average` is run with the old data set and the new average value is compared to the old one.
+
+## pytest
+
+- pytest is installed by: `pip install pytest`.
+- All tests can be run using the command-line tool called `pytest`.
+- **Comparing floating point variables** needs to be handled in functions like `find_average` and is done using `pytest.approx(value, abs)`.
+- Even if one test fails, pytest runs all the tests and gives a report on the failing test. The assertion failure report generated my pytest is also more detailed than the usual Python assertion report.
+- pytest can run unittest suites directly using `pytest tests`.
```

Comment:
This information is incomplete. The command `pytest tests` expects that there is a directory `tests`. If I go in the directory `examples/python-testing` and run `pytest tests` I just get an error:

```text
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.8.10, pytest-6.2.5, py-1.11.0, pluggy-0.13.0
rootdir: /home/jaustar/teaching/courses/simulation-software-engineering/lecture-material/05_testing_and_ci/examples/python-testing
collected 0 items                                                                                                                                                                                                 

============================================================================================== no tests ran in 0.00s ==============================================================================================
ERROR: file or directory not found: tests

```

Could you maybe give some example command to run that I can actually use on the example code provided by this repository?